microbench/src/main/java/io/netty/microbench/concurrent/DefaultPromiseBenchmark.java

```diff
@@ -0,0 +1,162 @@
+    private EventExecutor executor;
+    private TestGenericFutureListener[] listeners;
+    private TestGenericFutureListener[] lastListeners;
+    private static AtomicLong doneCount = new AtomicLong();
+    private static CountDownLatch testLatch;
+    private final Runnable toEventLoopRunnable = new Runnable() {
+        @Override
+        public void run() {
+            startOutEventLoop();
+        }
+    };
+
+    @Param({ "100000", "1000000" })
+    public int numPromises;
+
+    @Param({ "1", "4" })
+    public int numListeners;
+
+    @Param({ "0", "4" })
+    public int lateListeners;
+
+    @AuxCounters
+    @State(Scope.Thread)
+    public static class DoneCounter {
+        @Setup(Level.Iteration)
+        public void clean() {
+            doneCount.set(0);
+        }
+
+        public long doneCount() {
+            return doneCount.get();
+        }
+    }
+
+    @Setup(Level.Trial)
+    public void setupTrial() {
+        executorService = Executors.newSingleThreadExecutor();
+        executor = new DefaultEventExecutor(executorService);
+        listeners = new TestGenericFutureListener[numListeners];
+        for (int i = 0; i < listeners.length; ++i) {
+            listeners[i] = new TestGenericFutureListener();
+        }
+        lastListeners = new TestGenericFutureListener[lateListeners];
+        for (int i = 0; i < lastListeners.length; ++i) {
+            lastListeners[i] = new TestGenericFutureListener();
+        }
+    }
+
+    @Setup(Level.Invocation)
+    public void setupInvocation() {
+        testLatch = new CountDownLatch(numPromises * (listeners.length + lastListeners.length));
+    }
+
+    @TearDown(Level.Trial)
+    public void teardownTrial() {
+        final Future<?> f;
+        try {
+            executorService.shutdown();
+        } finally {
+            f = executor.shutdownGracefully(0, 0, TimeUnit.MILLISECONDS);
+        }
+        try {
+            executorService.awaitTermination(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+        try {
+            f.get(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        } catch (ExecutionException e) {
+            e.printStackTrace();
+        } catch (TimeoutException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Benchmark
+    public void testInEventLoop(DoneCounter doneCounter) throws Exception {
+        startInEventLoop();
+        testLatch.await();
+    }
+
+    @Benchmark
+    public void testOutEventLoop(DoneCounter doneCounter) throws Exception {
+        startOutEventLoop();
+        testLatch.await();
+    }
+
+    private void startInEventLoop() {
+        executor.execute(toEventLoopRunnable);
+    }
+
+    private void startOutEventLoop() {
+        for (int i = 0; i < numPromises; ++i) {
+            Promise<Void> p = new DefaultPromise<Void>(executor);
+            p.addListeners(listeners);
```

Comment:
> What are you trying to benchmark?

@nmittler asked the same question above  see my response https://github.com/netty/netty/pull/5222#discussion_r62549140.

> addListeners vs addListener
The goal is to add multiple listeners. The previous implementation of `DefaultPromise` synchronized in each iteration of the loop (instead of once outside the loop) so it is possible the new implementation will have an advantage for `addListeners`.  I could add methods to the benchmark which use `addListener` too.